add servo support to MAG_MOUNTED_PROBE

ellensp opened the pull request:
### Description

User built  machine with a MAG_MOUNTED_PROBE on a servo arm, a non supported configuration.
I added basic support for it. 

### Requirements

MAG_MOUNTED_PROBE
NUM_SERVOS
Z_SAFE_HOMING (a requirement) 
new 
MAG_MOUNTED_PROBE_SERVO_NR
MAG_MOUNTED_PROBE_SERVO_ANGLES

### Benefits

A  MAG_MOUNTED_PROBE on a servo arm has a chance of working.

User requesting this has provided videos that it work as intended


https://github.com/user-attachments/assets/2f4f3eef-89d2-4938-ab45-28b39178788e

ellensp committed:
add servo support to MAG_MOUNTED_PROBE

ellensp committed:
disable MAG_MOUNTED_PROBE_SERVO_NR by default

gjdodd commented:
I should have looked here first, just spent the afternoon implementing this myself. is there any chance of this branch been updated to latest, and also would it help get this one through if I helped test it out.

ellensp commented:
@gjdodd updated to current bugfix 2.1.x

gjdodd commented:
@ellensp, Thankyou for updating the branch. There is one small issue, currently enabling the MAG_MOUNTED_PROBE_SERVO_NR, will cause a compliation error, it needs the NUM_SERVOS block in "Conditionals-4-adv.h" updating to automatically increment the servo count. Like in the code below, once this is in place it will compile. Just going to start testing as I have switched everything over on my machine now, so hopefully it all goes well.

![image](https://github.com/user-attachments/assets/4e62b70b-7e8a-4ea5-a917-93a1edc96f06)

ellensp committed:
auto set NUM_SERVOS if omitted

ellensp commented:
I was unaware of that particular check,  I just always set NUM_SERVOS...

Added

gjdodd commented:
found one more slight issue, if you enable the deactivate servos after move option, it will fail to compile, line 983 in SanityCheck.h needs !defined(MAG_MOUNTED_PROBE_SERVO_NR) adding to the if clause.

So far everything else is looking good

gjdodd commented:
would you prefer me to create PR's into your branch instead of comments if I find anything else

gjdodd commented:
I have spotted an issue in the deploy/stow code, it relates to the new PROBE_TRIGGERED_WHEN_STOWED_TEST work by the looks of it, as the probe finishes the homing cycle it is running through the stow code so quick that the condition for PROBE_TRIGGERED == deploy does not work correctly as the condition for checking the probe state is not what would be expected, I am just looking at what to do to fix it, might be we need a stow delay similar to the servo delay or something with a config to enable it or not, or the other option is to ignore the trigger state when stowing and just continue to try and stow.

What are your thoughts @ellensp

gjdodd commented:
Scratch that, I have now seen what the issue is, needed to look at the printer a bit closer, at the end of the zero it is not moving the z away from the probe and so it is staying triggered so the condition check does not match and so the stow code never runs. Just looking at what to do with it now.

gjdodd committed:
- Fix issue with failing probe stow as position after z homing does not fire until after stowing has occured, should just continue to stow the probe as it has just finished homing
- Add pre deploy and stow steps to allow for position of nozzle to be in a safe place after ABL etc
- Include sanity check for deactivate servos to include mounted probe servo

gjdodd committed:
Merge branch 'add-servo-support-to-magnetic-probe' of https://github.com/ellensp/Marlin into ellensp/add-servo-support-to-magnetic-probe

gjdodd commented:
@ellensp I have created a PR on your repo to fix the stowing issues with not moving z before stowing which triggers errors with miss matched endpoint and deploy/stow flag.

It also includes a couple of extra configs to stop the nozzle crashing into the servo deployed arm after ABL etc

ellensp committed:
Merge pull request #3 from gjdodd/feature/add-servo-support-to-magnetic-probe

Updates and a couple of fixes for add servo support to magnetic probe

gjdodd commented:
I have been testing this for a couple of days now, and everything works as expected, things which have been tested

Z home
Linear bed leveling
Z align
Probe repeatability test

Thanks @ellensp

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27551

thinkyhead committed:
config tweaks

thinkyhead committed:
add test
